File: test/experimental/fex_rate.cpp
////////////////////////////////////////////////////////////////////////////////
// fex_rate.cpp
// ------------
//
// Optimal combined control of the exchange rate, as formulated in [1].
//
// [1] Mundaca, Gabriela, and Bernt Øksendal. "Optimal stochastic intervention
// control with application to the exchange rate." Journal of Mathematical
// Economics 29.2 (1998): 225-243.
//
// Author: Parsiad Azimzadeh
////////////////////////////////////////////////////////////////////////////////

#include <QuantPDE/Core>
#include <QuantPDE/Modules/HJBQVI>

using namespace QuantPDE;
using namespace QuantPDE::Modules;

////////////////////////////////////////////////////////////////////////////////

#include <cmath>  // fabs
#include <limits> // numeric_limits

using namespace std;

////////////////////////////////////////////////////////////////////////////////

int main() {

	// Dimensions
	constexpr int Dimension = 1;
	constexpr int StochasticControlDimension = 1;
	constexpr int ImpulseControlDimension = 1;

	// Boundaries
	const Real x_min = -6.;
	const Real x_max = +6.;

	// Number of points to use in discretizations
	const int points = 32;

	// Parameter that controls the clustering of nodes to the optimal parity
	const Real intensity = 10.;

	// Optimal parity
	const Real m = 0.;

	// Interest rate differential bounds
	const Real q_min = 0.;
	const Real q_max = 0.05;

	// Expiry time (infinity for corresponding elliptic problem)
	//const Real T = 10.;
	const Real T = numeric_limits<double>::infinity();

	// Discount
	const Real rho = 0.02;

	// Volatility
	const Real sigma = 0.3;

	// Other parameters
	const Real a = 0.25;
	const Real b = 3.;
	const Real lambda = 1.;
	const Real c = 0.;

	// Number of timesteps
	const int timesteps = 32;

	// How to handle the control
	auto method = HJBQVIControlMethod::FULLY_IMPLICIT;
	//auto method = HJBQVIControlMethod::FULLY_EXPLICIT;

	// Maximum level of refinement
	// Solution and control data are printed at this level of refinement
	const int max_refinement = 6;

	// x axis
	const Axis x = Axis::cluster(
		x_min,
		m,
		x_max,
		points,
		intensity
	);

	// Problem description
	HJBQVI<
		Dimension,
		StochasticControlDimension,
		ImpulseControlDimension
	> hjbqvi(
		// Initial spatial grid
		{ x },

		// Initial stochastic control grid
		{ Axis::uniform( q_min, q_max, points ) },

		// Initial impulse control grid
		{ x },

		// Expiry
		T,

		// Discount
		[=] (Real t, Real x) {
			return rho;
		},

		// Volatility
		{
			[=] (Real t, Real x) {
				return sigma;
			}
		},

		// Controlled drift
		{
			[=] (Real t, Real x, Real q) {
				return -a * q;
			}
		},

		// Uncontrolled drift
		{
			[=] (Real t, Real x) {
				return 0.;
			}
		},

		// Controlled continuous flow
		[=] (Real t, Real x, Real q) {
			return -q * q * b;
		},

		// Uncontrolled continuous flow
		[=] (Real t, Real x) {
			const Real tmp = max(x - m, 0.);
			return -tmp * tmp;
		},

		// Transition
		{
			[=] (Real t, Real x, Real x_new) {
				return x_new;
			}
		},

		// Impulse flow
		[=] (Real t, Real x, Real x_new) {
			const Real zeta = x_new - x;
			return - lambda * fabs(zeta) - c;
		},

		// Exit function
		[=] (Real t, Real x) {
			return 0.;
		},

		// Initial number of timesteps
		timesteps,

		// How to handle the control
		method,

		// Bounded domain?
		false,

		// Refine stochastic control?
		true,

		// Refine impulse control?
		true,

		// Are the coefficients time independent?
		true
	);

	// Run
	HJBQVI_main(
		hjbqvi,
		{ m }, // Convergence test at optimal parity
		max_refinement
	);

	return 0;

}

File: test/experimental/fex_rate.m
## Dimensions $d$
dimension = 1;
stochastic_control_dimension = 1;
impulse_control_dimension = 1;

## Expiry $T$
expiry = inf;

## Volatility $\sigma$
function res = volatility (t, x)
  res = [0.3];
end

## Drift $\mu$
function res = drift (t, x, q)
  a = 0.25;
  res = { [-a * q] , [0.] };
end

## Discount factor $\rho$
discount = 0.02;

## Continuous flow $f$
function res = continuous_flow (t, x, q)
  b = 3.;
  tmp = max(m - x, 0.);
  res = { [-q * q * b] , [-tmp * tmp] };
end

## Intervention transition function $\Gamma$
function res = transition (t, x, x_new)
  res = [x_new];
end

## Intervention flow $K$
function res = impulse_flow (t, x, x_new)
  c = 0.;
  lambda = 1.;
  zeta = x_new - x;
  res = -lambda * abs(zeta) - c;
end

## Flow $g$ upon exiting the parabolic (resp. elliptic for infinite-horizon
## problems) boundary $
function res = exit_flow (t, x)
  res = 0.;
end

## Spatial grid
spatial_grid = [-6 -4.96909 -4.11317 -3.40211 -2.81087 -2.31861 -1.90802   ...
		-1.56461 -1.27631 -1.03296 -0.825989 -0.648105 -0.493045   ...
		-0.355349 -0.230168 -0.113093 0 0.113093 0.230168 0.355349 ...
		0.493045 0.648105 0.825989 1.03296 1.27631 1.56461 1.90802 ...
		2.31861 2.81087 3.40211 4.11317 4.96909 6];

## Stochastic control grid
q_min = 0.; q_max = 0.05; dq = q_max - q_min / ( length(spatial_grid) - 1 );
stochastic_control_grid = q_min : dq : q_max;
clear q_min q_max dq;

## Impulse control grid
impulse_control_grid = spatial_grid;

## Use a Neumann boundary condition $D_x u = 0$
## Useful for unbounded problems with artificial boundaries
zero_neumann = true;

## Specifies whether or not to refine grids
## Useful for bang-bang controls
stochastic_control_grid_refine = true;
impulse_control_grid_refine = true;

